File: tools/mtrace/mtrace-reader.py
#!/usr/bin/env python3
# SPDX-License-Identifier: BSD-3-Clause
#
# Copyright (c) 2022, Intel Corporation. All rights reserved.

#pylint:disable=mixed-indentation

# Tool to stream data from Linux SOF driver "mtrace" debugfs
# interface to standard output. Plain "cat" is not sufficient
# as each read() syscall returns log data with a 32bit binary
# header, containing the payload length.

import struct
import os
import sys
import argparse

READ_BUFFER = 16384
MTRACE_FILE = "/sys/kernel/debug/sof/mtrace/core0"

parser = argparse.ArgumentParser()
parser.add_argument('-m', '--mark-chunks',
                    action='store_true')

args = parser.parse_args()

chunk_idx = 0

fd = os.open(MTRACE_FILE, os.O_RDONLY)
while fd >= 0:
    # direct unbuffered os.read() must be used to comply with
    # debugfs protocol used. each non-zero read will return
    # a buffer containing a 32bit header and a payload
    read_bytes = os.read(fd, READ_BUFFER)

    # handle end-of-file
    if len(read_bytes) == 0:
        continue

    if len(read_bytes) <= 4:
        continue

    header = struct.unpack('I', read_bytes[0:4])
    data_len = header[0]
    data = read_bytes[4:4+data_len]

    if (args.mark_chunks):
        chunk_msg = "\n--- Chunk #{} start (size: {}) ---\n" .format(chunk_idx, data_len)
        sys.stdout.write(chunk_msg)

    sys.stdout.buffer.write(data)
    sys.stdout.flush()
    chunk_idx += 1

File: tools/mtrace/mtrace-reader.c
// SPDX-License-Identifier: BSD-3-Clause
//
// Copyright(c) 2025 Intel Corporation.
//
// Author: Kai Vehmanen <kai.vehmanen@linux.intel.com>
//
// Compile instructions:
//   gcc -o mtrace-reader mtrace-reader.c -Wall -O2
//

#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <stdint.h>

#define READ_BUFFER 16384
#define MTRACE_FILE "/sys/kernel/debug/sof/mtrace/core0"

uint8_t buffer[READ_BUFFER];

int main(void)
{
	ssize_t read_bytes;
	uint32_t data_len;
	uint32_t header;
	int fd;

	fd = open(MTRACE_FILE, O_RDONLY);
	if (fd < 0) {
		perror("open");
		return 1;
	}

	while (1) {
		read_bytes = read(fd, buffer, READ_BUFFER);

		/* handle end-of-file */
		if (read_bytes == 0)
			continue;

		if (read_bytes <= 4)
			continue;

		header = *(uint32_t *)buffer;
		data_len = header;
		if (data_len > read_bytes - 4)
			continue;

		if (write(STDOUT_FILENO, buffer + 4, data_len) < 0) {
			perror("write");
			return -1;
		}
	}

	close(fd);

	return 0;
}
